source/fab/steps/compile_fortran.py

```diff
@@ -70,111 +91,212 @@ def run(self, artefact_store, config):
+        object_files = artefact_store.setdefault(OBJECT_FILES, defaultdict(set))
+        for root, source_files in build_trees.items():
+            new_objects = [lookup[af.fpath].output_fpath for af in source_files]
+            object_files[root].update(new_objects)
+
     # todo: identical to the c version - make a super class
-    def compile_file(self, analysed_file: AnalysedFile):
-        output_fpath = analysed_file.fpath.with_suffix('.o')
+    def process_file(self, analysed_file: AnalysedFile):
+        """
+        Prepare to compile a fortran file, and compile it if anything has changed since it was last compiled.
+
+        Returns a compilation result, including various hashes from the time of compilation:
+
+        * hash of the source file
+        * hash of the compile flags
+        * hash of the module files on which we depend
 
-        # already compiled?
-        if self._config.reuse_artefacts and output_fpath.exists():
+        """
+        # flags for this file
+        flags = self.flags.flags_for_path(
+            path=analysed_file.fpath, source_root=self._config.source_root,
+            project_workspace=self._config.project_workspace)
+        flags_hash = string_checksum(str(flags))
+
+        # do we need to recompile?
+        last_compile = self._last_compiles.get(analysed_file.fpath)
+        recompile_reasons = self.recompile_check(analysed_file, flags_hash, last_compile)
+
+        if recompile_reasons:
+            try:
+                logger.debug(f'CompileFortran {recompile_reasons} for {analysed_file.fpath}')
+                self.compile_file(analysed_file, flags)
+            except Exception as err:
+                return Exception("Error compiling file:", err)
+        else:
             log_or_dot(logger, f'CompileFortran skipping: {analysed_file.fpath}')
+
+        # get the hashes of the modules we depend on
+        # record them so we know if they've changed next time we compile.
+        module_deps_hashes = {}
+        for mod_dep in analysed_file.module_deps:
+            if mod_dep not in self._mod_hashes:
+                logger.debug(f"Dependecy {mod_dep} appears to be external to this project and hasn't been hashed.")
+                continue
+            module_deps_hashes[mod_dep] = self._mod_hashes[mod_dep]
+
+        return CompiledFile(
+            input_fpath=analysed_file.fpath, output_fpath=analysed_file.fpath.with_suffix('.o'),
+            source_hash=analysed_file.file_hash, flags_hash=flags_hash,
+            module_deps_hashes=module_deps_hashes
+        )
+
+    def recompile_check(self, analysed_file: AnalysedFile, flags_hash: int, last_compile: Optional[CompiledFile]):
+
+        # todo: other environmental considerations for the future:
+        #   - env vars, e.g OMPI_FC
+        #   - compiler version
+
+        recompile_reasons = []
+        # first encounter?
+        if not last_compile:
+            recompile_reasons.append(NO_PREVIOUS_RESULT)
```

Comment:
That's true. We chose to give the user _all_ the reasons instead of just the first, as it seemed more useful to them, and the time saving seems to be negligible. Happy to reconsider.